src/odemis/driver/piezomotor.py

```diff
@@ -0,0 +1,733 @@
+            else:
+                self._axis_map[axis_name] = axis_par['axis_number']
+
+            # WFM stepsize
+            try:
+                self._wfm_stepsizes[axis_name] = axis_par['wfm_stepsize']  # approximately 5e-6 m / step
+            except KeyError:
+                raise ValueError("Axis %s has no wfm stepsize." % axis_name)
+
+            # Axis range
+            try:
+                axis_range = axis_par['range']
+            except KeyError:
+                logging.info("Axis %s has no range. Assuming (-1, 1)", axis_name)
+                axis_range = (-1, 1)
+
+            # Axis unit
+            try:
+                axis_unit = axis_par['unit']
+            except KeyError:
+                axis_unit = "m"
+                logging.info("Axis %s has no unit. Assuming %s", axis_name, axis_unit)
+
+            # Axis speed
+            try:
+                self._speed = axis_par['speed']
+            except KeyError:
+                # m if linear, "rad" otherwise
+                self._speed = DEFAULT_AXIS_SPEED
+                logging.info("Axis %s was not given a speed value. Assuming %s", axis_name, self._speed)
+
+            # Axis mode (closed loop/ open loop)
+            try:
+                closed_loop = axis_par['closed_loop']
+            except KeyError:
+                closed_loop = True
+                logging.info("Axis mode (closed/open loop) not specified for axis %s. Assuming closed loop.", axis_name)
+            self._closed_loop[axis_name] = closed_loop
+
+            ad = model.Axis(canAbs=closed_loop, unit=axis_unit, range=axis_range)
+            axes_def[axis_name] = ad
+
+        Actuator.__init__(self, name, role, axes=axes, inverted=inverted, **kwargs)
+        self._executor = CancellableThreadPoolExecutor(max_workers=1)  # one task at a time
+        self._ser_access = threading.RLock()
+
+        # Connect to hardware
+        self._port = None  # port number
+        min_axis = min(self._axis_map.values())
+        self._serial = self._findDevice(port, min_axis)
+        self._recovering = False
+
+        # Get version
+        hwVersions = []
+        for ax_name, ax_num in self._axis_map.items():
+            ver = self.getVersion(ax_num)
+            sn = self.getSerialNumber(ax_num)
+            hwVersions.append("Axis %s ('%s') version: %s, " % (ax_num, ax_name, ver) +
+                              "serial number: %s" % sn)
+        self._hwVersion = ", ".join(hwVersions)
+
+        # Position and referenced VAs
+        self.position = model.VigilantAttribute({}, unit="m", readonly=True)
+        self.referenced = model.VigilantAttribute({}, unit="m", readonly=True)
+
+        # Configuration
+        for axis in self._axis_map.values():
+            self.setWaveform(axis, WAVEFORM_DELTA)
+
+        driver_name = getSerialDriver(self._port)
+        self._swVersion = "serial driver: %s" % (driver_name,)
+
```

Comment:
I thought we only need referencing for open-loop? Doesn't the encoder deal with the absolute position in closed-loop?

Reply:
That's kind of the opposite. Note that in theory, referencing is orthogonal to closed-/open-loop. Referencing  just means that you're able to have a fixed origin, by determining the position of a "reference switch" (or "index").
In practice, however, with open-loop this has little use because after a move, due to imprecision, this origin will shift randomly a little bit. That's why referencing is typically used *only* with closed-loop systems.